### PR TITLE
329 Update Sphinx Docs 'User Guide' --> 'Extend GeoIPS With Plugins'

### LINKED ISSUE #79: Document Interface and Plugin classes and update Plugin documentation
# Requested Update

### Description
After #78, we need to update documentation for Interfaces and Plugins.

### Background and Motivation
*  [ OPTIONAL: Provide additional detailed information and background related to this issue ]

### Alternative Solutions
*  [ OPTIONAL: If appropriate, list alternative fixes that could be considered]
  
### Environment
*  [ IF APPROPRIATE: Corresponding repository versions ]

### Code to demonstrate issue
*  [ IF APPROPRIATE: Can include steps here to identify/demonstrate bug fix or desired feature ]

### Checklist for Completion
* [ ]

### LINKED ISSUE #329: Update the 'User Guide' --> 'Extend GeoIPS with Plugins' section of the GeoIPS Sphinx Documentation
# Requested Update

### Description
The 'User Guide' --> 'Extend GeoIPS with Plugins' Section of the GeoIPS Sphinx documentation is entirely empty, asides from the list of available Module/YAML plugins. This section needs updated, and it will be done in a hierarchical fashion. That being, a top level section describing the overall nature of GeoIPS (ie. it's plugin based structure), and lower level sections, which describe the nature of each individual plugin, both module and YAML based. These sections should largely mimic the tutorial slides which we presented recently.

### Background and Motivation
This section will be great for new users and expert users, as it will have a series of documentation accurately describing each step of the GeoIPS development path. This will greatly improve peoples ability to create their own GeoIPS-based packages.

### PR DESCRIPTION
**Pull request by evrose54**
# Reviewer Checklist

* [x] Existing tests pass (integration_tests/full_test.sh, test_all.sh, pytests as appropriate.
      Includes linting checks and documentation build)

  * [x] No full testing required
---

* [x] Appropriate unit tests added and pass for new/modified functionality
      
  * [x] No unit tests required
---

* [x] Appropriate integration tests added and pass for new/modified functionality
      
  * [x] No integration tests required.
---

* [x] Appropriate documentation added for new/modified functionality
      
  * [x] No documentation required
---

* [x] Appropriate release notes added for new/modified functionality
      
  * [x] No release note updates required.
---

https://github.com/NRLMMD-GEOIPS/.github/blob/main/.github/review-template.md

# Related Issues
fixes #79
fixes #329 

# Summary
This PR adds documentation for everything we discussed in the tutorial.
Includes:
  - Products
  - Product Defaults
  - Algorithms
  - Readers
  - Static Sectors
  - Feature Annotators
  - Gridline Annotators
  - Colormappers
  - Output Formatters
  - And Updated Top Level Info for Extending GeoIPS with Plugins

### CODE REVIEW

**Comment by evrose54** [comment]
Haven't yet figured out how to create subdirectories for each plugin type in .rst. If either @jsolbrig or @mindyls know how to do this, please add a comment to this PR so I can implement this next week!

**Comment by evrose54** [comment]
Note: I haven't yet added to the 'release notes' since I'm not sure which release this will actually take place in. Since we just released v1.11.3, I'll wait until v1.11.4 is created (or v2.0.0), to add the changes to the docs.

File: docs/source/starter/installation.rst [hunk 1]
```diff
@@ -10,6 +10,8 @@
  | # # # for more details. If you did not receive the license, for more information see:
  | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
 
+.. _complete_install:
```
Comment (mindyls) [inline docs/source/starter/installation.rst:13]: This is named differently in the other "links" PR

**Review by mindyls — DISMISSED** [review]
Once the link naming is de-conflicted with the other open PR, I am good to go with this one.

File: docs/source/userguide/plugin_extend.rst [hunk 2]
```diff
@@ -1,21 +1,195 @@
...
+   | # # # Naval Research Laboratory, Marine Meteorology Division
+   | # # #
+   | # # # This program is free software: you can redistribute it and/or modify it under
+   | # # # the terms of the NRLMMD License included with this program. This program is
+   | # # # distributed WITHOUT ANY WARRANTY; without even the implied warranty of
+   | # # # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the included license
+   | # # # for more details. If you did not receive the license, for more information see:
+   | # # # https://github.com/U-S-NRL-Marine-Meteorology-Division/
+
+.. _plugin-extend:
```
Comment (mindyls) [inline docs/source/userguide/plugin_extend.rst:15]: This is awesome!

**Review by mindyls — COMMENTED** [review]
@jsolbrig @evrose54 @AJThorpe 
I"m thinking once @AJThorpe is done with datatree / pytest updates, it would probably be great for him to 1) build this documentation and 2) open up the html in his browser and try following along with extending GeoIPS with plugins.

He can provide feedback if anything does't make sense (since we are all too close to it at this point to necessarily notice...), and can identify any typos, etc in the RST links, etc.

@AJThorpe you can build the documentation with:
```
$GEOIPS_PACKAGES_DIR/geoips/docs/build_docs.sh $GEOIPS_PACKAGES_DIR/geoips geoips
```
Then you can just open up index.html in your browser:
```
$GEOIPS_PACKAGES_DIR/geoips/build/html/index.html
```
Just click through the documentation and make sure it all renders properly, and that it all makes sense, and you are able to set up your own plugin from scratch by following these instructions.

**Comment by AJThorpe** [comment]
@evrose54 @mindyls 

I keep getting this error throw when trying to build, it happens right after building the sphinx PDF documentation. I couldn't find a specific work around or solution online, but can post the full output if needed.
```
Warning, treated as error:
/home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:17:Duplicate explicit target name: "plugin-extend".
make: *** No targets specified and no makefile found.  Stop.
```

**Comment by mindyls** [comment]
> @evrose54 @mindyls 
> 
> I keep getting this error throw when trying to build, it happens right after building the sphinx PDF documentation. I couldn't find a specific work around or solution online, but can post the full output if needed.
> ```
> Warning, treated as error:
> /home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:17:Duplicate explicit target name: "plugin-extend".
> make: *** No targets specified and no makefile found.  Stop.
> ```

Grep the docs for plugin-extend, the reference is probably defined twice.  Probably just need to delete one of the refs

**Comment by AJThorpe** [comment]
> > @evrose54 @mindyls
> > I keep getting this error throw when trying to build, it happens right after building the sphinx PDF documentation. I couldn't find a specific work around or solution online, but can post the full output if needed.
> > ```
> > Warning, treated as error:
> > /home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:17:Duplicate explicit target name: "plugin-extend".
> > make: *** No targets specified and no makefile found.  Stop.
> > ```
> 
> Grep the docs for plugin-extend, the reference is probably defined twice. Probably just need to delete one of the refs. It may also be resolved by merging main into it (or that could be what caused the duplicate)

This leads to a new error when building the sphinx docs, should this link to another rst file or another section in the rst file?:
```
Warning, treated as error:
/home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:96:undefined label: 'complete_install'
```

**Comment by mindyls** [comment]
Take a look at this comment - I think 2 different PRs had differently named links:
https://github.com/NRLMMD-GEOIPS/geoips/pull/330#discussion_r1342022133

Probalby need to rename one or the other.

> > > @evrose54 @mindyls
> > > I keep getting this error throw when trying to build, it happens right after building the sphinx PDF documentation. I couldn't find a specific work around or solution online, but can post the full output if needed.
> > > ```
> > > Warning, treated as error:
> > > /home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:17:Duplicate explicit target name: "plugin-extend".
> > > make: *** No targets specified and no makefile found.  Stop.
> > > ```
> > 
> > 
> > Grep the docs for plugin-extend, the reference is probably defined twice. Probably just need to delete one of the refs. It may also be resolved by merging main into it (or that could be what caused the duplicate)
> 
> This leads to a new error when building the sphinx docs, should this link to another rst file or another section in the rst file?:
> 
> ```
> Warning, treated as error:
> /home/ajthorp/geoips/geoips/build/buildfrom_docs/source/userguide/plugin_extend.rst:96:undefined label: 'complete_install'
> ```

**Review by AJThorpe — APPROVED** [review]

File: docs/source/userguide/plugin_development/algorithm.rst [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+
+To extend GeoIPS with a new ``algorithm`` plugin, first follow the :ref:`instructions for
+setting up a plugin package<plugin-development-setup>`.
+
+Module plugins are required to have several top-level variables:
+    * name
+    * interface
+    * family
+    * docstring
+
```
Comment (mindyls) [inline docs/source/userguide/plugin_development/algorithm.rst:15]: @jsolbrig @AJThorpe @evrose54 Andrew said this documentation was much improved, and he could easily folow it to set up a new plugin.  Do you think we should get this merged as is, and we could open an additional issue to expand it to be more narrative style in the future?

Comment (mindyls) [reply docs/source/userguide/plugin_development/algorithm.rst:15]: Good to go.

**Review by jsolbrig — APPROVED** [review]

**Review by mindyls — APPROVED** [review]
Mindy will open issue to expand on this documentation and link here.
https://github.com/NRLMMD-GEOIPS/geoips/issues/398

Mindy will also open issue assigned to Jeremy to look through this documentation.
https://github.com/NRLMMD-GEOIPS/geoips/issues/399

Then resolve conflicts, make sure everything still builds, then merge.

**Pull request merged**